File: advent-of-code-2024/day3.py
with open('input') as f:
    cData = f.read().strip()

def part1(Data):
    Data = [mulOpr for mulOpr in Data.split('mul') if len(mulOpr) > 0 and mulOpr[0] == '(' and ')' in mulOpr]

    Data = [oprnd[1:].split(')')[0] for oprnd in Data if "," in oprnd]

    Data = [int(num.split(',')[0]) * int(num.split(',')[1]) for num in Data if num.split(',')[0].isnumeric() and  1 <= len(num.split(',')[0]) <= 3 and num.split(',')[1].isnumeric() and 1 <= len(num.split(',')[1]) <= 3]

    return sum(Data)

print("part1 : ", part1(cData))

cData = cData.split("don't()")

part2 = part1(cData[0])

for i in range(1, len(cData)):
    if 'do()' in cData[i]:
        for eachDo in cData[i].split('do()')[1:]:
            part2 += part1(eachDo)

print("part2 : ", part2)

File: advent-of-code-2024/cpp/day3.cpp
# include <iostream>
# include <fstream>

using namespace std;

long long checkMul(string & inp, int i){

    if(inp[i+1] == 'u' && inp[i+2] == 'l' && inp[i+3] == '('){
        int n1 = 0, n2 = 0;
        i += 4;

        while(isnumber(inp[i])) n1 = n1 * 10 + (inp[i++] - '0');
        if(inp[i] == ',') i += 1;
        while(isnumber(inp[i])) n2 = n2 * 10 + (inp[i++] - '0');

        if(inp[i] == ')') return n1 * n2;
    }

    return 0;
}

int main(){
    ifstream file("../input");

    char ch; 
    string inp="";

    while(file >> ch){
        inp += ch;
    }

    long long part1 = 0, part2 = 0;    

    bool do_ = true;

    for(int i=0; i<inp.size(); i++){
        if(inp[i] == 'm') part1 += checkMul(inp, i);

        if(do_ && inp[i] == 'm') part2 += checkMul(inp, i);

        if(do_ && i + 6 < inp.length() && inp.substr(i, 7) == "don't()") do_ = false;

        if(!do_ && i+3 < inp.length() && inp.substr(i, 4) == "do()") do_ = true;
    }

    cout << "part1 : " << part1 << endl;
    cout << "part2 : " << part2 << endl;
}